src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs

```diff
@@ -239,6 +211,42 @@ private static ProcessModuleCollection GetModules(int processId, bool firstModul
                 }
             }
         }
+
+        private static void EnumProcessModulesUntilSuccess(SafeProcessHandle processHandle, IntPtr[]? modules, int size, out int needed)
+        {
+            // EnumProcessModules is not a reliable method to get the modules for a process.
+            // If OS loader is touching module information, this method might fail and copy part of the data.
+            // This is no easy solution to this problem. The only reliable way to fix this is to
+            // suspend all the threads in target process. Of course we don't want to do this in Process class.
+            // So we just to try avoid the race by calling the same method 50 (an arbitrary number) times.
+            for (int i = 0; i < 50; i++)
+            {
+                if (Interop.Kernel32.EnumProcessModules(processHandle, modules, size, out needed))
+                {
+                    return;
+                }
+
+                Thread.Sleep(1);
+            }
+
+            throw new Win32Exception();
```

Comment:
This may pick up the last error from the `Thread.Sleep` call instead of the `EnumProcessModules` call.